Add warning for JAX versions on import of Dynamics

DanPuzzuoli opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Closes #231 

This PR updates the JAX type registration in `dispatch`, and also adds a warning at JAX import time about the version bounds for JAX.

### Details and comments

The docs build has also been set to run with JAX version 0.4.3 to avoid this warning.

wshanks commented on qiskit_dynamics/dispatch/backends/jax.py:
```diff
@@ -17,26 +17,29 @@
+        from jax.interpreters.xla import DeviceArray
 
-        JAX_TYPES += (DynamicJaxprTracer,)
+        JAX_TYPES += (DeviceArray,)
```
I don't know jax well enough to judge -- is it okay to drop the extra tracer types in this PR?

DanPuzzuoli replied:
Yeah I'm not 100% sure but with the arraylias work this appears to be fine (also all the Dynamics tests passed with this change). They used to have a zoo of tracer and abstract array types but in more recent version they've introduced the `Array` and `Tracer`, which i'm assuming the other types inherit from.

DanPuzzuoli replied:
See the latest comment on the PR, but I've actually removed `DeviceArray` completely now. Only `Array` and `Tracer` are required for any JAX version above `0.4.0`.

wshanks commented on qiskit_dynamics/dispatch/backends/jax.py:
```diff
@@ -17,26 +17,29 @@
+            "The functionality in the perturbation module of Qiskit Dynamics requires a JAX "
+            "version <= 0.4.6, due to a bug in JAX versions > 0.4.6. For versions 0.4.4, 0.4.5, "
+            "and 0.4.6, using the perturbation module functionality requires setting "
+            "os.environ['JAX_JIT_PJIT_API_MERGE'] = '0' before importing JAX."
```
Perhaps you could check `os.environ` and not warn if this is set?

DanPuzzuoli replied:
Ah that's a really good idea, I'll make this change.

wshanks commented on tox.ini:
```diff
@@ -41,8 +41,8 @@ commands = black {posargs} qiskit_dynamics test
     -r{toxinidir}/requirements-dev.txt
-    jax<=0.4.6
-    jaxlib<=0.4.6
+    jax<=0.4.3
```
You might want a comment here to explain the lower pin. Or maybe you don't need the lower pin if you follow my suggestion of checking the environment variable before warning.

DanPuzzuoli replied:
I'll implement your suggestion, making this moot.

wshanks reviewed (dismissed):
I had a couple suggestions but this looks fine to me (other than the tracer type changes which I am not sure about).

DanPuzzuoli commented:
@wshanks I've implemented your suggestion for the warning to not trigger if the JAX version is 0.4.4, 0.4.5, or 0.4.6 and they've already set the `os` environment variable. I've reverted the docs to use JAX 0.4.6 as the warning will no longer be triggered.

DanPuzzuoli commented:
So I still ended up needing to set `os.environ["JAX_JIT_PJIT_API_MERGE"] = "0"` in  `docs/conf.py` to avoid the warning in the docs build, but luckily that worked. I've put that in with a comment, and have added a description of this PR to issue #190 (where I'm keeping track of all the little changes that we've had to make because of these JAX issues).

I was actually also able to completely remove adding `DeviceArray` to `JAX_TYPES` - `Array` and `Tracer` turn out to be all that's necessary these days.

wshanks reviewed (approved):
Looks good!
